server/plugin.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"sync"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron"
+)
+
+// Plugin implements the interface expected by the Mattermost server to communicate between the server and plugin processes.
+type Plugin struct {
+	plugin.MattermostPlugin
+
+	// configurationLock synchronizes access to the configuration.
+	configurationLock sync.RWMutex
+
+	// configuration is the active plugin configuration. Consult getConfiguration and
+	// setConfiguration for usage.
+	configuration *configuration
+
+	// a job for pre-calculating channel recommendations for users.
+	preCalcJob    *cron.Cron
+	preCalcPeriod string
+	botUserID     string
+}
+
+// ServeHTTP demonstrates a plugin that handles HTTP requests by greeting the world.
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	fmt.Fprint(w, "Hello, world!")
+}
+
+// See https://developers.mattermost.com/extend/plugins/server/reference/
+
+type readFile func(filename string) ([]byte, error)
+
+func (p *Plugin) setupBot(reader readFile) error {
+	botID, err := p.Helpers.EnsureBot(&model.Bot{
+		Username:    "suggestions",
+		DisplayName: "Suggestions",
+		Description: "Created by the Suggestions plugin.",
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to ensure github bot")
+	}
+	p.botUserID = botID
+	bundlePath, err := p.API.GetBundlePath()
+	if err != nil {
+		return errors.Wrap(err, "couldn't get bundle path")
+	}
+
+	profileImage, err := reader(filepath.Join(bundlePath, "assets", "profile.jpeg"))
+	if err != nil {
+		return errors.Wrap(err, "couldn't read profile image")
+	}
+
+	appErr := p.API.SetProfileImage(botID, profileImage)
+	if appErr != nil {
+		return errors.Wrap(appErr, "couldn't set profile image")
+	}
+	return nil
+}
+
+func (p *Plugin) startPrecalcJob() error {
+	config := p.getConfiguration()
+	p.preCalcPeriod = "@daily" // default once a day
+	if config.PreCalculationPeriod != "" {
+		p.preCalcPeriod = config.PreCalculationPeriod
+	}
+	c := cron.New()
+	if err := c.AddFunc(p.preCalcPeriod, func() {
+		p.preCalculateRecommendations()
+	}); err != nil {
+		return err
+	}
+	c.Start()
+	p.preCalcJob = c
+	return nil
+}
+
+// OnActivate will be run on plugin activation.
+func (p *Plugin) OnActivate() error {
+	p.API.RegisterCommand(getCommand())
+	err := p.initStore()
+	if err != nil {
+		return err
+	}
+	err = p.setupBot(ioutil.ReadFile)
+	if err != nil {
+		return err
+	}
+	err = p.startPrecalcJob()
+	if err != nil {
+		return err
+	}
+	go p.preCalculateRecommendations() //Run pre-calculation at once
```

Comment:
I have certain concerns about the performance in a real system with for example 5000 users. Looks like the approach of measure all the activity of all the users during the whole history plus the KNN algorithm running for each user, for me sounds super heavy weight. Running this at the beginning looks the easy approach, but sounds risky if this is installed in a crowded installation.

Reply:
Yeap, this is a concern. Precalculations can be run once a day(or any other period of time). The very first precalcuation is the heaviest, because it pulls all posts from the database, aggregates data and counts scores of the channels per user. Next precalcualtions pull only new data from the database. In future aggregated data should be saved in the DB to eliminate this work.

Apart from that there is a KNN algorithm which is run on every precalculation. Its complexity is O(N*M*M), where N is number of users, M is number of channels. This algorithm is also quite heavy, but 5000 users will not be a problem. There is an alternative here. We can calculate user suggestion on demand, which will take O(M*M). In this case user experience might not be acceptable so I preferred first version.

Reply:
In practice, I think `N >> M` in most cases, so it seems like calculating user suggestions on demand might actually be very efficient.

I've got a couple of ideas for how we might be able to make performance acceptable from the get go, but we can chat tomorrow :)